test/e2e/volume_provisioning.go

```diff
@@ -108,61 +216,182 @@ var _ = framework.KubeDescribe("Dynamic Provisioning", func() {
 	})
 
 	framework.KubeDescribe("DynamicProvisioner", func() {
-		It("should create and delete persistent volumes [Slow] [Volume]", func() {
-			framework.SkipUnlessProviderIs("openstack", "gce", "aws", "gke")
-
-			By("creating a StorageClass")
-			class := newStorageClass("", "internal")
-			class, err := c.StorageV1().StorageClasses().Create(class)
-			defer c.StorageV1().StorageClasses().Delete(class.Name, nil)
-			Expect(err).NotTo(HaveOccurred())
-
-			By("creating a claim with a dynamic provisioning annotation")
-			claim := newClaim(ns)
-			claim.Spec.StorageClassName = &class.Name
-
-			defer func() {
-				c.Core().PersistentVolumeClaims(ns).Delete(claim.Name, nil)
-			}()
-			claim, err = c.Core().PersistentVolumeClaims(ns).Create(claim)
-			Expect(err).NotTo(HaveOccurred())
-
-			if framework.ProviderIs("vsphere") {
-				// vsphere provider does not allocate volumes in 1GiB chunks, so setting expected size
-				// equal to requestedSize
-				testDynamicProvisioning(c, claim, requestedSize)
-			} else {
-				// Expected size of the volume is 2GiB, because the other three supported cloud
-				// providers allocate volumes in 1GiB chunks.
-				testDynamicProvisioning(c, claim, "2Gi")
-			}
-		})
-	})
+		// This test checks that dynamic provisioning can provision a volume
+		// that can be used to persist data among pods.
+
+		tests := []storageClassTest{
+			{
+				"should provision SSD PD on GCE/GKE",
+				[]string{"gce", "gke"},
+				"kubernetes.io/gce-pd",
+				map[string]string{
+					"type": "pd-ssd",
+					// Check that GCE can parse "zone" parameter, however
+					// we can't create PDs in different than default zone
+					// as we don't know if we're running with Multizone=true
+					"zone": framework.TestContext.CloudConfig.Zone,
+				},
+				"1.5Gi",
+				"2Gi",
+				func(volume *v1.PersistentVolume) error {
+					return checkGCEPD(volume, "pd-ssd")
+				},
+			},
+			{
+				"should provision HDD PD on GCE/GKE",
+				[]string{"gce", "gke"},
+				"kubernetes.io/gce-pd",
+				map[string]string{
+					"type": "pd-standard",
+				},
+				"1.5Gi",
+				"2Gi",
+				func(volume *v1.PersistentVolume) error {
+					return checkGCEPD(volume, "pd-standard")
+				},
+			},
+			// AWS
+			{
+				"should provision gp2 EBS on AWS",
```

Comment:
as we think about out-of-tree provisioning, does it seem better to move the various cloud tests to their own file?